File: app/Sources/Parley/VAD.swift
import Foundation

// Voice-activity detection over mic buffers. Pure math + a small state machine so
// it can be unit-tested with synthetic sample buffers (no audio device).

enum Level {
    /// Root-mean-square of float samples in [-1, 1].
    static func rms(_ samples: [Float]) -> Float {
        guard !samples.isEmpty else { return 0 }
        var sum: Float = 0
        for s in samples { sum += s * s }
        return (sum / Float(samples.count)).squareRoot()
    }

    /// dBFS. Silence (rms 0) is clamped to a floor instead of -inf.
    static func dB(_ rms: Float, floor: Float = -120) -> Float {
        guard rms > 0 else { return floor }
        return max(floor, 20 * log10(rms))
    }
}

/// ADAPTIVE silence VAD. A fixed threshold cut people off during quiet-but-present speech
/// and natural pauses. Instead we track a running average of the user's OWN speaking
/// loudness (EWMA) and only end once the level drops FAR below that average (a big relative
/// drop) AND stays there for `trailingSilence`. Soft speakers get a low bar, loud speakers
/// a high one; a mid-sentence breath never ends the turn.
struct SilenceVAD {
    var trailingSilence: Double = 1.8         // sustained strong drop before ending
    var dropDB: Float = 22                    // "extreme drop" = this far below the average

    private var speechAvgDB: Float = -32
    private var hasAvg = false
    private(set) var started = false
    private(set) var silenceElapsed: Double = 0

    init(trailingSilence: Double = 1.8, dropDB: Float = 22) {
        self.trailingSilence = trailingSilence
        self.dropDB = dropDB
    }

    enum Decision: Equatable {
        case waiting    // pre-speech silence, keep listening
        case speaking   // active speech
        case ended      // trailing silence exceeded -> stop recording
    }

    mutating func process(rmsDB: Float, duration: Double) -> Decision {
        // Dynamic floor relative to how loud the user has been speaking (clamped sane).
        let threshold = min(-45, max(-70, speechAvgDB - dropDB))
        if rmsDB >= threshold {
            started = true
            silenceElapsed = 0
            if rmsDB > -50 {                  // only real speech feeds the average
                if hasAvg { speechAvgDB = speechAvgDB * 0.95 + rmsDB * 0.05 }
                else { speechAvgDB = rmsDB; hasAvg = true }
            }
            return .speaking
        }
        guard started else { return .waiting }
        silenceElapsed += duration
        return silenceElapsed >= trailingSilence ? .ended : .speaking
    }

    mutating func reset() {
        started = false
        silenceElapsed = 0
        speechAvgDB = -32
        hasAvg = false
    }
}

File: windows/Vad.cs
namespace Parley;

// Silence-timer VAD — direct port of the macOS SilenceVAD (same thresholds).
public sealed class Vad
{
    public float SpeechThresholdDb { get; init; } = -50f;
    public double TrailingSilence { get; init; } = 0.9;

    public bool Started { get; private set; }
    private double _silenceElapsed;

    public enum Decision { Waiting, Speaking, Ended }

    public Decision Process(float rmsDb, double duration)
    {
        if (rmsDb >= SpeechThresholdDb)
        {
            Started = true;
            _silenceElapsed = 0;
            return Decision.Speaking;
        }
        if (!Started) return Decision.Waiting;
        _silenceElapsed += duration;
        return _silenceElapsed >= TrailingSilence ? Decision.Ended : Decision.Speaking;
    }

    public void Reset()
    {
        Started = false;
        _silenceElapsed = 0;
    }

    public static float Rms(ReadOnlySpan<float> samples)
    {
        if (samples.Length == 0) return 0;
        double sum = 0;
        foreach (var s in samples) sum += s * s;
        return (float)Math.Sqrt(sum / samples.Length);
    }

    public static float Db(float rms) => rms > 0 ? MathF.Max(-120f, 20f * MathF.Log10(rms)) : -120f;
}

public static class Wav
{
    /// 16 kHz mono 16-bit PCM → WAV bytes (44-byte header + data).
    public static byte[] Encode(ReadOnlySpan<short> samples, int sampleRate = 16000)
    {
        var dataLen = samples.Length * 2;
        using var ms = new MemoryStream(44 + dataLen);
        using var w = new BinaryWriter(ms);
        w.Write("RIFF"u8); w.Write(36 + dataLen); w.Write("WAVE"u8);
        w.Write("fmt "u8); w.Write(16); w.Write((short)1); w.Write((short)1);
        w.Write(sampleRate); w.Write(sampleRate * 2); w.Write((short)2); w.Write((short)16);
        w.Write("data"u8); w.Write(dataLen);
        foreach (var s in samples) w.Write(s);
        return ms.ToArray();
    }
}
